### PR TITLE
WIP: Removal of nlohmann::json

### PR DESCRIPTION
**Pull request by Macroshock**
Putting this here so people can look at this. The purpose is to replace all usage of nlohmann::json with rapidjson. There are some changes that I'm not sure if they were done right, such as the ApiDispatcher::publicKeysToAddresses method or the new JsonException in JsonHelper.h, so this needs to be reviewed and also tested.

### CODE REVIEW

File: src/WalletApi/ApiDispatcher.h [hunk 1]
```diff
@@ -326,9 +320,9 @@ class ApiDispatcher
         /* END OF API FUNCTIONS */
         //////////////////////////
 
-        /* Extracts {host, port, ssl, filename, password}, from body */
```
Comment (zpalmtree) [inline src/WalletApi/ApiDispatcher.h]: Comment got nuked

File: src/WalletApi/ApiDispatcher.cpp [hunk 2]
```diff
@@ -307,27 +313,22 @@ void ApiDispatcher::middleware(
...
+            res.set_content(std::string(string_buffer.GetString()) + "\n", "application/json");
         }
         else
         {
             res.status = statusCode;
         }
     }
-    /* Most likely a key was missing. Do the error handling here to make the
-       rest of the code simpler */
-    catch (const json::exception &e)
```
Comment (zpalmtree) [inline src/WalletApi/ApiDispatcher.cpp]: Should we not be catching `JsonException` here?

File: src/Nigel/Nigel.cpp [hunk 3]
```diff
@@ -377,62 +411,84 @@ std::tuple<bool, std::vector<CryptoNote::RandomOuts>> Nigel::getRandomOutsByAmou
...
+    for (uint64_t item : amounts) {
+        writer.Uint64(item);
+    }
+    writer.EndArray();
 
     /* The blockchain cache doesn't call it outs_count
        it calls it mixin */
-    if (m_isBlockchainCache)
-    {
-        j.erase("outs_count");
```
[the remark this answers is not in the document]
Comment (Macroshock) [reply src/Nigel/Nigel.cpp]: I don't recall doing this on purpose so it's a mistake.

File: src/WalletApi/ApiDispatcher.cpp [hunk 4]
```diff
@@ -910,35 +955,45 @@ std::tuple<Error, uint16_t> ApiDispatcher::setNodeInfo(
...
+	writer.String(daemonHost);
+    writer.Key("daemonPort");
+	writer.Uint(daemonPort);
+    writer.Key("daemonSSL");
+	writer.Bool(daemonSSL);
+    writer.Key("nodeFee");
+	writer.Uint64(nodeFee);
+    writer.Key("nodeAddress");
+	writer.String(nodeAddress);
+	writer.EndObject();
```
[the remark this answers is not in the document]
Comment (Macroshock) [reply src/WalletApi/ApiDispatcher.cpp]: This indentation was not intentional, I think it was caused by VS2017 so I'll get rid of it.

File: src/WalletApi/ApiDispatcher.cpp [hunk 5]
```diff
@@ -1456,43 +1644,56 @@ std::tuple<Error, uint16_t> ApiDispatcher::getBalanceForAddress(
...
     const auto balances = m_walletBackend->getBalances();
 
-    nlohmann::json j;
-
+    rapidjson::StringBuffer string_buffer;
+    rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(string_buffer);
+    writer.StartObject();
+    /* I had to assign a key to the array. While nlohmann allows a nameless array as a unique member of
+    an object, rapidjson apparently doesn't like it.*/
+    writer.Key("balances");
```
[the remark this answers is not in the document]
Comment (Macroshock) [reply src/WalletApi/ApiDispatcher.cpp]: I tried doing this in a small project I used to test rapidjson, and it didn't work because it failed an assertion. Will try again later to see if I made a mistake.

File: src/Nigel/Nigel.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Macroshock) [reply src/Nigel/Nigel.cpp]: Ok I just remembered why I did this, what the original code did was it created a json object with both the amounts array and the outs_count and then if the if statement resolved to true it would erase the outs_count and write the mixin. 

Now, before that If statement the only thing that has been written is the amounts array, when the If resolves to true, it writes the mixin and if it resolves to false, it writes the outs_count. 
I thought by doing this I could achieve the same thing.

File: src/WalletApi/ApiDispatcher.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Macroshock) [reply src/WalletApi/ApiDispatcher.cpp]: I did some testing and realized that the problem is that it should be just a nameless array, not a nameless array contained inside an object. This will be fixed soon.

File: include/CryptoNote.h [hunk 6]
```diff
@@ -29,6 +28,36 @@ struct KeyInput {
...
+
+    writer.Key("k_image");
+    keyImage.toJSON(writer);
+    
+    writer.EndObject();
+  }
+
+  /* Initializes the class from a json value */
+  void fromJSON(const JSONValue &j) {
+    amount = getUintFromJSON(j, "amount");
```
[the remark this answers is not in the document]
Comment (brandonlehmann) [reply include/CryptoNote.h]: Anywhere we touch a uint64_t it should be get `getUint64FromJSON`

File: include/CryptoTypes.h [hunk 7]
```diff
@@ -1,19 +1,8 @@
 // Copyright (c) 2012-2017, The CryptoNote developers, The Bytecoin developers
+// Copyright (c) 2014-2018, The Monero Project
```
Comment (brandonlehmann) [inline include/CryptoTypes.h]: What did we add from Monero here?

File: include/WalletTypes.h [hunk 8]
```diff
@@ -25,6 +25,32 @@ namespace WalletTypes
...
+
+        /* Initializes the class from a json value */
+        void fromJSON(const JSONValue &j) {
+            key.fromString(getStringFromJSON(j, "key"));
+            amount = getUint64FromJSON(j, "amount");
+            /* If we're talking to a daemon or blockchain cache
+            that returns the globalIndex as part of the structure
+            of a key output, then we need to load that into the
+            data structure. */
+            if (j.HasMember("globalIndex"))
```
Comment (brandonlehmann) [inline include/WalletTypes.h]: Please put braces in here

File: include/WalletTypes.h [hunk 9]
```diff
@@ -49,6 +75,51 @@ namespace WalletTypes
...
         uint64_t unlockTime;
+
+        /* Converts the class to a json object */
+        void toJSON(rapidjson::Writer<rapidjson::StringBuffer> &writer) const
+        {
+            writer.StartObject();
+
+            writer.Key("outputs");
+            writer.StartArray();
+            for (const auto &item : keyOutputs) {
```
Comment (brandonlehmann) [inline include/WalletTypes.h]: Style change, move the open brace to newline

File: include/WalletTypes.h [hunk 10]
```diff
@@ -25,6 +25,32 @@ namespace WalletTypes
...
+            key.toJSON(writer);
+
+            writer.Key("amount");
+            writer.Uint64(amount);
+
+            writer.EndObject();
+        }
+
+        /* Initializes the class from a json value */
+        void fromJSON(const JSONValue &j) {
```
Comment (brandonlehmann) [inline include/WalletTypes.h]: Style change, move the open brace to newline

File: include/CryptoNote.h [hunk 11]
```diff
@@ -29,6 +28,36 @@ struct KeyInput {
...
+    for (const auto &item : getArrayFromJSON(j, "key_offsets")) {
```
Comment (brandonlehmann) [inline include/CryptoNote.h]: Style change, move the open brace to newline

File: include/CryptoNote.h [hunk 12]
```diff
@@ -29,6 +28,36 @@ struct KeyInput {
...
+    writer.EndArray();
+
+    writer.Key("k_image");
+    keyImage.toJSON(writer);
+    
+    writer.EndObject();
+  }
+
+  /* Initializes the class from a json value */
+  void fromJSON(const JSONValue &j) {
```
Comment (brandonlehmann) [inline include/CryptoNote.h]: Style change, move the open brace to newline

File: include/JsonHelper.h [hunk 13]
```diff
@@ -11,6 +11,19 @@ typedef rapidjson::GenericObject<true, rapidjson::GenericValue<rapidjson::UTF8<c
 
 typedef rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator>> JSONValue;
 
+class JsonException : public std::exception {
+	std::string message;
+
+    public:
+        JsonException(std::string err) {
```
Comment (brandonlehmann) [inline include/JsonHelper.h]: Style change, move the open brace to newline

File: include/JsonHelper.h [hunk 14]
```diff
@@ -11,6 +11,19 @@ typedef rapidjson::GenericObject<true, rapidjson::GenericValue<rapidjson::UTF8<c
...
+            message = err;
+        }
+
+        const char* what () const throw () {
```
Comment (brandonlehmann) [inline include/JsonHelper.h]: Style change, move the open brace to newline

File: include/JsonHelper.h [hunk 15]
```diff
@@ -11,6 +11,19 @@ typedef rapidjson::GenericObject<true, rapidjson::GenericValue<rapidjson::UTF8<c
 
 typedef rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator>> JSONValue;
 
+class JsonException : public std::exception {
```
Comment (brandonlehmann) [inline include/JsonHelper.h]: Style change, move the open brace to newline

File: include/WalletTypes.h [hunk 16]
```diff
@@ -49,6 +75,51 @@ namespace WalletTypes
...
+            transactionPublicKey.toJSON(writer);
+
+            writer.Key("unlockTime");
+            writer.Uint64(unlockTime);
+
+            writer.EndObject();
+        }
+
+        /* Initializes the class from a json value */
+        void fromJSON(const JSONValue &j) {
```
Comment (brandonlehmann) [inline include/WalletTypes.h]: Style change, move the open brace to newline

File: include/WalletTypes.h [hunk 17]
```diff
@@ -49,6 +75,51 @@ namespace WalletTypes
...
+            keyOutputs.clear();
+            for (const auto &x : getArrayFromJSON(j, "outputs")) {
```
Comment (brandonlehmann) [inline include/WalletTypes.h]: Style change, move the open brace to newline

File: include/WalletTypes.h [hunk 18]
```diff
@@ -49,6 +75,51 @@ namespace WalletTypes
...
+                KeyOutput keyOutput;
+                keyOutput.fromJSON(x);
+                keyOutputs.push_back(keyOutput);
+            }
+            hash.fromString(getStringFromJSON(j, "hash"));
+            transactionPublicKey.fromString(getStringFromJSON(j, "txPublicKey"));
+            /* We need to try to get the unlockTime from an integer in the json
```
Comment (brandonlehmann) [inline include/WalletTypes.h]: Insert space here to break up the wall of text pls

File: include/WalletTypes.h [hunk 19]
```diff
@@ -49,6 +75,51 @@ namespace WalletTypes
...
+            however, if that fails because we're talking to a blockchain
+            cache API that encodes unlockTime as a string (due to json
+            integer encoding limits), we need to attempt this as a string */
+            try {
```
Comment (brandonlehmann) [inline include/WalletTypes.h]: Style change, move the open brace to newline

File: include/WalletTypes.h [hunk 20]
```diff
@@ -60,6 +131,73 @@ namespace WalletTypes
...
         std::vector<CryptoNote::KeyInput> keyInputs;
+
+        /* Converts the class to a json object */
+        void toJSON(rapidjson::Writer<rapidjson::StringBuffer> &writer) const
+        {
+            writer.StartObject();
+
+            writer.Key("outputs");
+            writer.StartArray();
+            for (const auto &item : keyOutputs) {
```
Comment (brandonlehmann) [inline include/WalletTypes.h]: Style change, move the open brace to newline

File: include/WalletTypes.h [hunk 21]
```diff
@@ -60,6 +131,73 @@ namespace WalletTypes
...
+
+            writer.Key("unlockTime");
+            writer.Uint64(unlockTime);
+
+            writer.Key("paymentID");
+            writer.String(paymentID);
+
+            writer.Key("inputs");
+            writer.StartArray();
+            for (const auto &item : keyInputs) {
```
Comment (brandonlehmann) [inline include/WalletTypes.h]: Style change, move the open brace to newline

File: include/WalletTypes.h [hunk 22]
```diff
@@ -60,6 +131,73 @@ namespace WalletTypes
...
+            writer.EndArray();
+
+            writer.EndObject();
+        }
+
+        /* Initializes the class from a json value */
+        void fromJSON(const JSONValue &j)
+        {
+            keyOutputs.clear();
+            for (const auto &x : getArrayFromJSON(j, "outputs")) {
```
Comment (brandonlehmann) [inline include/WalletTypes.h]: Style change, move the open brace to newline

File: include/WalletTypes.h [hunk 23]
```diff
@@ -60,6 +131,73 @@ namespace WalletTypes
...
+            }
+
+            hash.fromString(getStringFromJSON(j, "hash"));
+            transactionPublicKey.fromString(getStringFromJSON(j, "txPublicKey"));
+            unlockTime = getUint64FromJSON(j, "unlockTime");
+            /* We need to try to get the unlockTime from an integer in the json
+            however, if that fails because we're talking to a blockchain
+            cache API that encodes unlockTime as a string (due to json
+            integer encoding limits), we need to attempt this as a string */
+            try {
```
Comment (brandonlehmann) [inline include/WalletTypes.h]: Style change, move the open brace to newline

File: include/WalletTypes.h [hunk 24]
```diff
@@ -60,6 +131,73 @@ namespace WalletTypes
...
+            for (const auto &x : getArrayFromJSON(j, "outputs")) {
+                KeyOutput keyOutput;
+                keyOutput.fromJSON(x);
+                keyOutputs.push_back(keyOutput);
+            }
+
+            hash.fromString(getStringFromJSON(j, "hash"));
+            transactionPublicKey.fromString(getStringFromJSON(j, "txPublicKey"));
+            unlockTime = getUint64FromJSON(j, "unlockTime");
+            /* We need to try to get the unlockTime from an integer in the json
```
Comment (brandonlehmann) [inline include/WalletTypes.h]: insert a new line to break up this wall of text

File: include/WalletTypes.h [hunk 25]
```diff
@@ -60,6 +131,73 @@ namespace WalletTypes
...
+            however, if that fails because we're talking to a blockchain
+            cache API that encodes unlockTime as a string (due to json
+            integer encoding limits), we need to attempt this as a string */
+            try {
+                unlockTime = getUint64FromJSON(j, "unlockTime");
+            } catch (const JsonException &e) {
```
Comment (brandonlehmann) [inline include/WalletTypes.h]: Style change, move the open brace to newline

File: include/WalletTypes.h [hunk 26]
```diff
@@ -60,6 +131,73 @@ namespace WalletTypes
...
+                unlockTime = std::stoull(getStringFromJSON(j, "unlockTime"));
+            }
+
+            paymentID = getStringFromJSON(j, "paymentID");
+
+            keyInputs.clear();
+            for (const auto &x : getArrayFromJSON(j, "outputs")) {
```
Comment (brandonlehmann) [inline include/WalletTypes.h]: Style change, move the open brace to newline

File: include/WalletTypes.h [hunk 27]
```diff
@@ -79,6 +217,49 @@ namespace WalletTypes
...
+        /* Converts the class to a json object */
+        void toJSON(rapidjson::Writer<rapidjson::StringBuffer> &writer) const
+        {
+            writer.StartObject();
+            writer.Key("coinbaseTX");
+            coinbaseTransaction.toJSON(writer);
+
+            writer.Key("transactions");
+            writer.StartArray();
+            for (const auto &tx : transactions) {
```
Comment (brandonlehmann) [inline include/WalletTypes.h]: Style change, move the open brace to newline

File: include/WalletTypes.h [hunk 28]
```diff
@@ -79,6 +217,49 @@ namespace WalletTypes
...
+            writer.EndObject();
+        }
+
+        /* Initializes the class from a json value */
+        void fromJSON(const JSONValue &j)
+        {
+            coinbaseTransaction.fromJSON(getJsonValue(j, "transactions"));
+
+            transactions.clear();
+            for (const auto &item : getArrayFromJSON(j, "transfers")) {
```
Comment (brandonlehmann) [inline include/WalletTypes.h]: Style change, move the open brace to newline

File: src/Miner/BlockchainMonitor.cpp [hunk 29]
```diff
@@ -81,13 +84,19 @@ void BlockchainMonitor::stop()
...
-    json j = {
-        {"jsonrpc", "2.0"},
-        {"method", "getlastblockheader"},
-        {"params", {}}
-    };
+    rapidjson::StringBuffer string_buffer;
+    rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(string_buffer);
 
-    auto res = m_httpClient->Post("/json_rpc", j.dump(), "application/json");
+    writer.StartObject();
```
Comment (brandonlehmann) [inline src/Miner/BlockchainMonitor.cpp]: There's got to be a better way to handle these... perhaps a struct with a json method instead? How this is done here obscures the the actual JSON structure :(

File: src/Miner/BlockchainMonitor.cpp [hunk 30]
```diff
@@ -109,35 +118,34 @@ std::optional<Crypto::Hash> BlockchainMonitor::requestLastBlockHash()
         return std::nullopt;
     }
 
-    try
-    {
-        json j = json::parse(res->body);
+    std::cout << "befure parse blockchain monitor" << std::endl;
+    try {
```
Comment (brandonlehmann) [inline src/Miner/BlockchainMonitor.cpp]: Style change, move the open brace to newline

File: src/Miner/BlockchainMonitor.cpp [hunk 31]
```diff
@@ -109,35 +118,34 @@ std::optional<Crypto::Hash> BlockchainMonitor::requestLastBlockHash()
...
         }
 
-        return j.at("result").at("block_header").at("hash").get<Crypto::Hash>();
-    }
-    catch (const json::exception &e)
-    {
+        Crypto::Hash hash;
+        hash.fromString(getStringFromJSON(getJsonValue(getJsonValue(j, "result"), "block_header"), "hash"));
+        return hash;    
+    } catch (const JsonException &e) {
```
Comment (brandonlehmann) [inline src/Miner/BlockchainMonitor.cpp]: Style change, move the open brace to newline

File: src/Miner/MinerManager.cpp [hunk 32]
```diff
@@ -228,15 +231,24 @@ void MinerManager::stopBlockchainMonitoring()
...
 
-    json j = {
-        {"jsonrpc", "2.0"},
-        {"method", "submitblock"},
-        {"params", {Common::toHex(toBinaryArray(minedBlock))}}
-    };
+    rapidjson::StringBuffer string_buffer;
+    rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(string_buffer);
+
+    writer.StartObject();
```
Comment (brandonlehmann) [inline src/Miner/MinerManager.cpp]: Same as above -- let's use a struct or something here to format this out and actually show, in the code, what the structure looks like

File: src/Miner/MinerManager.cpp [hunk 33]
```diff
@@ -256,16 +268,24 @@ BlockMiningParameters MinerManager::requestMiningParameters()
...
-                {"wallet_address", m_config.miningAddress},
-                {"reserve_size", 0}
-            }}
-        };
-
-        auto res = m_httpClient->Post("/json_rpc", j.dump(), "application/json");
+        rapidjson::StringBuffer string_buffer;
+        rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(string_buffer);
+
+        writer.StartObject();
```
Comment (brandonlehmann) [inline src/Miner/MinerManager.cpp]: Rinse & repeat above

File: src/Miner/MinerManager.cpp [hunk 34]
```diff
@@ -289,51 +309,43 @@ BlockMiningParameters MinerManager::requestMiningParameters()
             continue;
         }
 
-        try
-        {
-            json j = json::parse(res->body);
+        try {
```
Comment (brandonlehmann) [inline src/Miner/MinerManager.cpp]: Style change, move the open brace to newline

File: src/Miner/MinerManager.cpp [hunk 35]
```diff
@@ -289,51 +309,43 @@ BlockMiningParameters MinerManager::requestMiningParameters()
...
-
                 std::this_thread::sleep_for(std::chrono::seconds(1));
                 continue;
             }
-
             return params;
-        }
-        catch (const json::exception &e)
-        {
+        } catch (const JsonException &e) {
```
Comment (brandonlehmann) [inline src/Miner/MinerManager.cpp]: Style change, move the open brace to newline

File: src/Miner/MinerManager.cpp [hunk 36]
```diff
@@ -289,51 +309,43 @@ BlockMiningParameters MinerManager::requestMiningParameters()
...
+            params.difficulty = getUint64FromJSON(getJsonValue(j, "result"), "difficulty");
 
             std::vector<uint8_t> blob = Common::fromHex(
-                j.at("result").at("blocktemplate_blob").get<std::string>()
+                getStringFromJSON(getJsonValue(j, "result"), "blocktemplate_blob")
             );
 
-            if(!fromBinaryArray(params.blockTemplate, blob))
-            {
+            if(!fromBinaryArray(params.blockTemplate, blob)) {
```
Comment (brandonlehmann) [inline src/Miner/MinerManager.cpp]: Style change, move the open brace to newline

File: src/Rpc/CoreRpcServerCommandsDefinitions.h [hunk 37]
```diff
@@ -206,6 +206,26 @@ struct OutputEntry {
...
+    writer.Uint(global_amount_index);
+
+    writer.Key("out_key");
+    out_key.toJSON(writer);
+
+    writer.EndObject();
+  }
+
+  /* Initializes the class from a json value */
+  void fromJSON(const JSONValue &j) {
```
Comment (brandonlehmann) [inline src/Rpc/CoreRpcServerCommandsDefinitions.h]: Style change, move the open brace to newline

File: src/Rpc/CoreRpcServerCommandsDefinitions.h [hunk 38]
```diff
@@ -217,35 +237,36 @@ struct RandomOuts {
...
-inline void to_json(nlohmann::json &j, const OutputEntry &o)
-{
-    j = {
-        {"global_amount_index", o.global_amount_index},
-        {"out_key", o.out_key}
-    };
-}
+    writer.Key("outs");
+    writer.StartArray();
+    for (const auto& item : outs) {
```
Comment (brandonlehmann) [inline src/Rpc/CoreRpcServerCommandsDefinitions.h]: Style change, move the open brace to newline

File: src/Rpc/CoreRpcServerCommandsDefinitions.h [hunk 39]
```diff
@@ -217,35 +237,36 @@ struct RandomOuts {
...
-inline void from_json(const nlohmann::json &j, OutputEntry &o)
-{
-    o.global_amount_index = j.at("global_amount_index").get<uint32_t>();
-    o.out_key = j.at("out_key").get<Crypto::PublicKey>();
-}
+    writer.EndObject();
+  }
+
+  /* Initializes the class from a json value */
+  void fromJSON(const JSONValue &j) {
```
Comment (brandonlehmann) [inline src/Rpc/CoreRpcServerCommandsDefinitions.h]: Style change, move the open brace to newline

File: src/Rpc/CoreRpcServerCommandsDefinitions.h [hunk 40]
```diff
@@ -217,35 +237,36 @@ struct RandomOuts {
...
+    amount = getUint64FromJSON(j, "amount");
+    outs.clear();
+    for (const auto &x : getArrayFromJSON(j, "outputs")) {
```
Comment (brandonlehmann) [inline src/Rpc/CoreRpcServerCommandsDefinitions.h]: Style change, move the open brace to newline

File: src/WalletApi/ApiDispatcher.cpp [hunk 41]
```diff
@@ -230,20 +234,19 @@ void ApiDispatcher::middleware(
...
-        body = json::parse(req.body);
-        std::cout << "Body:\n" << std::setw(4) << body << std::endl;
-    }
-    /* Not neccessarily an error if body isn't needed */
-    catch (const json::exception &)
-    {
+    rapidjson::Document body;
+    
+    body.Parse(req.body);
+    if(!body.HasParseError()) {
```
Comment (brandonlehmann) [inline src/WalletApi/ApiDispatcher.cpp]: Style change, move the open brace to newline

File: src/WalletApi/ApiDispatcher.cpp [hunk 42]
```diff
@@ -230,20 +234,19 @@ void ApiDispatcher::middleware(
...
+        rapidjson::StringBuffer buffer;
+        rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+        body.Accept(writer);
+        std::cout << "Body:\n" << std::setw(4) << buffer.GetString() << std::endl;
+    } else {
```
Comment (brandonlehmann) [inline src/WalletApi/ApiDispatcher.cpp]:
```c
}
else
{
```

File: src/WalletApi/ApiDispatcher.cpp [hunk 43]
```diff
@@ -566,19 +580,22 @@ std::tuple<Error, uint16_t> ApiDispatcher::importViewAddress(
...
 
-    nlohmann::json j {
-        {"address", address}
-    };
+    rapidjson::StringBuffer string_buffer;
+	rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(string_buffer);
+	writer.StartObject();
+	writer.Key("address");
+	writer.String(address);
+	writer.EndObject();
```
[the remark this answers is not in the document]
Comment (brandonlehmann) [reply src/WalletApi/ApiDispatcher.cpp]: I do and will

File: src/WalletApi/ApiDispatcher.cpp [hunk 44]
```diff
@@ -680,35 +711,38 @@ std::tuple<Error, uint16_t> ApiDispatcher::sendAdvancedTransaction(
...
     {
         fee = getJsonValue<uint64_t>(body, "fee");
     }
 
     std::vector<std::string> subWalletsToTakeFrom = {};
 
-    if (body.find("sourceAddresses") != body.end())
-    {
-        subWalletsToTakeFrom = getJsonValue<std::vector<std::string>>(body, "sourceAddresses");
+    if (body.HasMember("sourceAddresses")) {
```
Comment (brandonlehmann) [inline src/WalletApi/ApiDispatcher.cpp]: Style change, move the open brace to newline

File: src/WalletApi/ApiDispatcher.cpp [hunk 45]
```diff
@@ -680,35 +711,38 @@ std::tuple<Error, uint16_t> ApiDispatcher::sendAdvancedTransaction(
...
+        for (const auto &item : getArrayFromJSON(body, "sourceAddresses")) {
```
Comment (brandonlehmann) [inline src/WalletApi/ApiDispatcher.cpp]: Style change, move the open brace to newline

File: src/WalletApi/ApiDispatcher.cpp [hunk 46]
```diff
@@ -776,9 +816,12 @@ std::tuple<Error, uint16_t> ApiDispatcher::sendAdvancedFusionTransaction(
 
     std::vector<std::string> subWalletsToTakeFrom;
 
-    if (body.find("sourceAddresses") != body.end())
-    {
-        subWalletsToTakeFrom = getJsonValue<std::vector<std::string>>(body, "sourceAddresses");
+	if (body.HasMember("sourceAddresses")) {
```
Comment (brandonlehmann) [inline src/WalletApi/ApiDispatcher.cpp]: Style change, move the open brace to newline

File: src/Nigel/Nigel.cpp [hunk 47]
```diff
@@ -192,47 +220,50 @@ bool Nigel::getDaemonInfo()
...
+            if(j.HasParseError())
+            {
+                throw JsonException(rapidjson::GetParseError_En(j.GetParseError()));
+            }
 
+            m_localDaemonBlockCount = getUint64FromJSON(j, "height");
             /* Height returned is one more than the current height - but we
-               don't want to overflow is the height returned is zero */
-            if (m_localDaemonBlockCount != 0)
+            don't want to overflow is the height returned is zero */
```
Comment (brandonlehmann) [inline src/Nigel/Nigel.cpp]: why did you remove the space here? put it back to keep formatting -- same with below

File: src/Nigel/Nigel.cpp [hunk 48]
```diff
@@ -192,47 +220,50 @@ bool Nigel::getDaemonInfo()
...
+            if (m_localDaemonBlockCount != 0)
```
Comment (brandonlehmann) [inline src/Nigel/Nigel.cpp]: remove errant spaces

File: src/Nigel/Nigel.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (brandonlehmann) [reply src/Nigel/Nigel.cpp]: @zpalmtree The cache will handle it w/o a problem; however, I don't care to send data we don't need to send. It's just wasteful TBH.

It may be better to put `outs_count` or `mixin` in a `std::string` instead and use the variable in the creation versus removing/deleting/adding/etc

File: src/Nigel/Nigel.cpp [hunk 49]
```diff
@@ -377,62 +411,84 @@ std::tuple<bool, std::vector<CryptoNote::RandomOuts>> Nigel::getRandomOutsByAmou
...
-    else
-    {
+    } else {
+        writer.Key("outs_count");
+        writer.Uint64(requestedOuts);
+        writer.EndObject();
+
         auto res = m_nodeClient->Post(
-            "/getrandom_outs", j.dump(), "application/json"
+            "/randomOutputs", string_buffer.GetString(), "application/json"
```
[the remark this answers is not in the document]
Comment (brandonlehmann) [reply src/Nigel/Nigel.cpp]: Shouldn't be. That only applies as daemon vs cache

File: src/WalletApi/ApiDispatcher.cpp [hunk 50]
```diff
@@ -230,20 +234,19 @@ void ApiDispatcher::middleware(
...
-    catch (const json::exception &)
-    {
+    rapidjson::Document body;
+    
+    body.Parse(req.body);
+    if(!body.HasParseError()) {
+        rapidjson::StringBuffer buffer;
+        rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+        body.Accept(writer);
+        std::cout << "Body:\n" << std::setw(4) << buffer.GetString() << std::endl;
```
[the remark this answers is not in the document]
Comment (brandonlehmann) [reply src/WalletApi/ApiDispatcher.cpp]: It's a debug statement anyways which should be wrapped or removed for production, correct?

**Comment by brandonlehmann** [comment]
@Macroshock Are you planning on circling back to this?

**Comment by Macroshock** [comment]
> @Macroshock Are you planning on circling back to this?

Sorry for not replying earlier, been busy with college and personal stuff. This week I'll resume working on this.

**Comment by brandonlehmann** [comment]
> > @Macroshock Are you planning on circling back to this?
> 
> Sorry for not replying earlier, been busy with college and personal stuff. This week I'll resume working on this.

No problem. Just making sure as I think we’re excited to see it wrapped up and merged.

File: src/Miner/BlockchainMonitor.cpp (hunk 29 above)
Comment (Macroshock) [reply src/Miner/BlockchainMonitor.cpp]: You mean to create a struct with a method that returns the JSON string that will be sent in the http post?

File: src/WalletApi/ApiDispatcher.cpp (hunk 50 above)
[the remark this answers is not in the document]
Comment (Macroshock) [reply src/WalletApi/ApiDispatcher.cpp]: Not sure that's a debug statement, I think it's intentional to display the response every time it receives one.

[the remark this answers is not in the document]
Comment (zpalmtree) [reply src/WalletApi/ApiDispatcher.cpp]: Yes, it's intentional.

**Comment by brandonlehmann** [comment]
@Macroshock This PR has been untouched for 30 days. If you plan on revisiting it to make it ready for inclusion, please let us know. It may be reopened at any time.
